shared/utility/math/filter/ParticleFilter.hpp

```diff
@@ -38,242 +72,396 @@ namespace utility::math::filter {
+            set_state(mean, covariance);
+            resample_method = residual_method_;
+
+            // The residual resampling method requires a secondary method to resample the residual particles. This
+            // secondary method cannot be the residual method.
+            if (residual_method == ResampleMethod::RESIDUAL) {
+                throw std::runtime_error(
+                    "Invalid setting for residual method. Can't use the residual resampling method to sample "
+                    "residuals.");
             }
-
-            return new_particles;
+            residual_method = resample_method_;
         }
 
-        ParticleList sample_particles(const StateVec& mean, const StateVec& covariance, const int& n_particles) {
-            // Sample single gaussian (represented by a gaussian mixture model of size 1)
-
-            // Implementation based on the work presented in
-            // Conrad Sanderson and Ryan Curtin.
-            // An Open Source C++ Implementation of Multi-Threaded Gaussian Mixture Models, k-Means and
-            // Expectation Maximisation. International Conference on Signal Processing and Communication
-            // Systems, 2017.
-
-            ParticleList new_particles =
-                ParticleList::NullaryExpr(Model::size, n_particles, [&]() { return norm(rng); });
+        /**
+         * @brief Constructs the particle filter with the provided values. The resampling method with default to the
+         * residual method with the systematic method for resampling the residual particles. It is an error to use the
+         * residual method for resampling residuals and an exception will be thrown.
+         *
+         * This version initialises the particles using multiple initial hypotheses. The particle count is split evenly
+         * across all hypotheses. If the division between the number of hypotheses and the number of particles is not
+         * even, then the remaining particles will be treated as rogues. It is an error for the numbers of means and
+         * covariances to differ.
+         *
+         * @param mean The mean of the multivariate normal distribution. Used for sampling the initial particles in the
+         * filter. There is one mean for each initial hypothesis.
+         * @param covariance The covariance of the multivariate normal distribution. Used for sampling the initial
+         * particles in the filter.  There is one covariance for each initial hypothesis.
+         * @param resample_method_ The method to use when resampling particles during the time update step.
+         * @param residual_method_ When using ResampleMethod::RESIDUAL for the resample method, this method will be used
+         * when the residual method needs to resample residual particles.
+         */
+        ParticleFilter(const std::vector<StateVec>& mean,
+                       const std::vector<StateMat>& covariance,
```

Comment:
If they need to be paired, then make them paired
```suggestion
        ParticleFilter(const std::vector<std::pair<StateVec, StateMat>>& pairs,
```
Never rely on the user for anything!

Reply:
Oh yea, I forgot that was a thing lol